File: datastruct/queue/LinkedListQueue.java
public class LinkedListQueue<E> implements Queue<E> {
    private class Node{
        public E e;
        public Node next;

        public Node(){
            next = null;
            e = null;
        }

        public Node(E e){
            this.e = e;
            this.next = null;
        }

        public Node(E e, Node next){
            this.e = e;
            this.next = next;
        }

        @Override
        public String toString(){
            return e.toString();
        }
    }

    private Node head, tail;
    private int size;

    public LinkedListQueue(){
        head = null;
        tail = null;
        size = 0;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public int getSize() {
        return size;
    }

    @Override
    public void enqueue(E ele) {
        if(tail == null){
            tail = new Node(ele);
            head = tail;
        }
        else{
            tail.next = new Node(ele);
            tail = tail.next;
        }
        size++;
    }

    @Override
    public E dequeue() {
        if (isEmpty()) {
            throw new IllegalArgumentException("Cannot dequeue from an empty queue!");
        }

        Node ret = head;
        head = head.next;
        ret.next = null;

        if (head == null) {
            tail = null;
        }
        size--;
        return ret.e;
    }

    @Override
    public E getFront() {
        if (isEmpty()) {
            throw new IllegalArgumentException("Cannot get element from an empty queue!");
        }

        return head.e;
    }

    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        str.append("Queue: front ");

        Node cur = head;
        while (cur != null) {
            str.append(cur + "->");
            cur = cur.next;
        }
        str.append("null tail");

        return str.toString();
    }
}

File: datastruct/queue/LinkedListQueue.py
from Queue import Queue

class Node(object):
    def __init__(self, ele = None, next_node = None):
        self.data = ele
        self.next = next_node
    
    def __str__(self):
        return str(self.data)

class LinkedListQueue(Queue):
    def __init__(self):
        self.__head = None
        self.__tail = None
        self.__size = 0
    
    def is_empty(self):
        return self.__size == 0
    
    def get_size(self):
        return self.__size

    def enqueue(self, ele):
        if self.__tail == None:
            self.__tail = Node(ele)
            self.__head = self.__tail
        else:
            self.__tail.next = Node(ele)
            self.__tail = self.__tail.next
        self.__size += 1
    
    def dequeue(self):
        if self.is_empty():
            raise IndexError('Cannot dequeue from an empty queue!')

        cur = self.__head
        self.__head = self.__head.next
        cur.next = Node

        if self.__head == None:
            self.__tail = None
        
        self.__size -= 1

        return cur.data
    
    def get_front(self):
        return self.__head.data
    
    def __str__(self):
        string = 'front '
        cur = self.__head

        while cur != None:
            string += str(cur) + '->'
            cur = cur.next
        string += 'null tail'
        return string

    

